### PR TITLE
file_upload.rs: sanitize path input

### LINKED ISSUE #466: Cannot upload to a symlink directory or any of subfolders
To reproduce, create a symbolic link to a directory and serve:

`
$ ln -s /xxx/xxx symlink
` 
`
$ miniserve -u ./
`

Then when trying to upload to the linked directory or any of its subdirectories, the upload fails with the follwing message:

> 400 Bad Request
> Invalid HTTP request
> caused by: Invalid value for 'path' parameter

Although the 'path' parameter in the url is valid and the file listing for it is correct:

` http://localhost:8080/upload?path=/symlink `

### LINKED ISSUE #518: 🚨 Potential Path Traversal (CWE-22)
👋 Hello, @svenstaro - a potential high severity Path Traversal (CWE-22) vulnerability in your repository has been disclosed to us.

#### Next Steps

1️⃣  Visit **https://huntr.dev/bounties/1-other-svenstaro/miniserve** for more advisory information.

2️⃣  **[Sign-up](https://huntr.dev/)** to validate or speak to the researcher for more assistance.

3️⃣ Propose a patch or outsource it to our community - whoever fixes it gets paid.

---

#### Confused or need more help?

- Join us on our **[Discord](https://huntr.dev/discord)** and a member of our team will be happy to help! 🤗

- Speak to a member of our team: @JamieSlome

---

*This issue was automatically generated by [huntr.dev](https://huntr.dev) - a bug bounty board for securing open source code.*

### PR DESCRIPTION
**Pull request by aliemjay**
Closes #518 
Closes #466 

Supersedes #476

### CODE REVIEW

File: tests/upload_files.rs [hunk 1]
```diff
@@ -78,3 +79,81 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
 }
+
+#[rstest]
+#[case("foo", "bar", "foo/bar")]
+#[case("/../foo", "bar", "foo/bar")]
+#[case("/foo", "/../bar", "foo/bar")]
+#[case("C:/foo", "C:/bar", if cfg!(windows) { "foo/bar" } else { "C:/foo/C:/bar" })]
+#[case(r"C:\foo", r"C:\bar", if cfg!(windows) { "foo/bar" } else { r"C:\foo/C:\bar" })]
+#[case(r"\foo", r"\..\bar", if cfg!(windows) { "foo/bar" } else { r"\foo/\..\bar" })]
+fn path_traversal(
```
Comment (svenstaro) [inline tests/upload_files.rs:90]: I really like the tests you added but please provide a doc comment and better test name so we know what it's supposed to test. :)

File: tests/upload_files.rs [hunk 2]
```diff
@@ -78,3 +79,81 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
 
     Ok(())
 }
+
+#[rstest]
+#[case("foo", "bar", "foo/bar")]
+#[case("/../foo", "bar", "foo/bar")]
+#[case("/foo", "/../bar", "foo/bar")]
+#[case("C:/foo", "C:/bar", if cfg!(windows) { "foo/bar" } else { "C:/foo/C:/bar" })]
```
Comment (svenstaro) [inline tests/upload_files.rs:87]: Good solution! Good to be thorough even on Windows.

File: src/file_upload.rs [hunk 3]
```diff
@@ -113,3 +118,30 @@ pub async fn upload_file(
         .append_header((header::LOCATION, return_path))
         .finish())
 }
+
+/// Guarantee that the path is relative and cannot traverse back to parent directories
+/// and optionally prevent traversing hidden directories.
```
Comment (svenstaro) [inline src/file_upload.rs:123]: For security critical stuff like this I often like an illustrating example in the documentation itself to make it entirely clear what kind of attacks are prevented by a given measure. As such, could you please add such an example here? Like which paths are input and what the expected output of this function is.

File: tests/upload_files.rs [hunk 4]
```diff
@@ -78,3 +79,81 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+    use std::os::windows::fs::symlink_dir;
+
+    // create symlink directory "foo" to point outside the root
+    let (dir, filename) = ("foo", "bar");
+    symlink_dir(tmpdir.path(), server.path().join(dir)).unwrap();
+
+    let full_path = server.path().join(dir).join(filename);
+    assert!(!full_path.exists());
+
+    // try to upload
```
Comment (svenstaro) [inline tests/upload_files.rs:142]: Super tiny nick: Please be consistent about comment capitalization. :)

File: src/file_upload.rs [hunk 5]
```diff
@@ -37,7 +37,7 @@ async fn save_file(
 /// Create new future to handle file as multipart data.
```
Comment (svenstaro) [inline src/file_upload.rs:37]: While not technically part of this PR (and we can do this in a follow-up), any chance you could write a better doc comment here? `save_file()` also has a pretty poor doc comment right now. :/

**Review by svenstaro — CHANGES_REQUESTED** [review]
Really solid PR and fixing an important security issue too! Just some nits. :)

File: tests/upload_files.rs [hunk 6]
```diff
@@ -78,3 +79,85 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+    // Make sure that the file was uploaded to the expected path
+    assert!(expected_path.exists());
+
+    Ok(())
+}
+
+#[rstest]
+#[case(server(&["-u"]), true)]
+#[case(server_no_stderr(&["-u", "--no-symlinks"]), false)]
+fn symlink(#[case] server: TestServer, #[case] ok: bool, tmpdir: TempDir) -> Result<(), Error> {
```
Comment (svenstaro) [inline tests/upload_files.rs:133]: This test could still use a doc comment and a better name.

**Comment by aliemjay** [comment]
3 typos in a single line. Nice :)

**Comment by aliemjay** [comment]
@svenstaro you may want to enable "auto cancel" option in travis-ci. It is still building previous commits before the forced-push.

**Comment by svenstaro** [comment]
> @svenstaro you may want to enable "auto cancel" option in travis-ci. It is still building previous commits before the forced-push.

Well, this is GitHub Actions and not Travis and I actually like it building every commit, even those that "don't matter". GitHub Actions is usually quick enough that it doesn't really matter.

**Review by svenstaro — APPROVED** [review]

**Comment by svenstaro** [comment]
Very cool!

**Comment by svenstaro** [comment]
@aliemjay there was actually a _very_ small bounty on this on huntr.dev for this. Apparently I'm about to receive said bounty but it feels wrong to collect this bounty for something that someone else did so please contact me via mail if you want that transferred somehow.

**Comment by aliemjay** [comment]
> it feels wrong to collect this bounty for something that someone else did so please contact me via mail if you want that transferred somehow.

No, I'm totally fine with this. Thanks for asking!

**Pull request merged**